superset/connectors/sqla/models.py

```diff
@@ -1011,7 +1013,7 @@ def adhoc_column_to_sqla(  # pylint: disable=too-many-locals
         if is_dttm and has_timegrain:
             sqla_column = self.db_engine_spec.get_timestamp_expr(
                 col=sqla_column,
-                pdf=None,
+                pdf=pdf,
```

Comment:
I'm concerned that this change may actually break existing logic—given it was explicitly set to `None`. Would you mind adding a unit test for this which helps not just to provide code coverage, but also helps reviewers et al. grok the consequence of the change.

Reply:
Unfortunately, in my testing whenever I tried to create a chart or use a dashboard, if a column was marked as `temporal` it would always call `get_timestamp_expr` via `adhoc_column_to_sqla` which means that the user defined date format is never passed to the DB Engine Spec.

It's possible that the root cause of the issue is that `get_timestamp_expr` is being called through `adhoc_column_to_sqla` which it should be getting called via `TableColumn.get_timestamp_expression` (the only other call path to `get_timestamp_expr` I could find. But all my tests pointed to `adhoc_column_to_sqla` being the root cause.

Reply:
@john-bodley the "pdf" is a shortcut for "date format (seconds or milliseconds)", this code was existing in many years, the "pdf" only used in Calculated Column but not used in Adhoc expression, so we shouldn't make this change.

<img width="895" alt="image" src="https://github.com/apache/superset/assets/2016594/4b6547bb-796e-4dd7-a2d5-07fcaff8b0a5">

Reply:
@ege-st 

The design of current Pinot DB spec is completely incorrect. Maintaining our own Pinot driver and db_spec should solve your issue. 


```python
class PinotEngineSpec(BaseEngineSpec):  # pylint: disable=abstract-method
    engine = "pinot"
    engine_name = "Apache Pinot"
    allows_subqueries = False
    allows_joins = False
    allows_alias_in_select = True
    allows_alias_in_orderby = False

    # https://docs.pinot.apache.org/users/user-guide-query/supported-transformations#datetime-functions
    _time_grain_expressions = {
        None: "{col}",
        "PT1S": "CAST(DATE_TRUNC('second', CAST({col} AS TIMESTAMP)) AS TIMESTAMP)",
        "PT1M": "CAST(DATE_TRUNC('minute', CAST({col} AS TIMESTAMP)) AS TIMESTAMP)",
        "PT5M": "CAST(ROUND(DATE_TRUNC('minute', CAST({col} AS TIMESTAMP)), 300000) as TIMESTAMP)",
        "PT10M": "CAST(ROUND(DATE_TRUNC('minute', CAST({col} AS TIMESTAMP)), 600000) as TIMESTAMP)",
        "PT15M": "CAST(ROUND(DATE_TRUNC('minute', CAST({col} AS TIMESTAMP)), 900000) as TIMESTAMP)",
        "PT30M": "CAST(ROUND(DATE_TRUNC('minute', CAST({col} AS TIMESTAMP)), 1800000) as TIMESTAMP)",
        "PT1H": "CAST(DATE_TRUNC('hour', CAST({col} AS TIMESTAMP)) AS TIMESTAMP)",
        "P1D": "CAST(DATE_TRUNC('day', CAST({col} AS TIMESTAMP)) AS TIMESTAMP)",
        "P1W": "CAST(DATE_TRUNC('week', CAST({col} AS TIMESTAMP)) AS TIMESTAMP)",
        "P1M": "CAST(DATE_TRUNC('month', CAST({col} AS TIMESTAMP)) AS TIMESTAMP)",
        "P3M": "CAST(DATE_TRUNC('quarter', CAST({col} AS TIMESTAMP)) AS TIMESTAMP)",
        "P1Y": "CAST(DATE_TRUNC('year', CAST({col} AS TIMESTAMP)) AS TIMESTAMP)",
    }

    @classmethod
    def column_datatype_to_string(
        cls, sqla_column_type: TypeEngine, dialect: Dialect
    ) -> str:
        # Pinot driver infers TIMESTAMP column as LONG, so make the quick fix.
        # When the Pinot driver fix this bug, current method could be removed.
        if isinstance(sqla_column_type, types.TIMESTAMP):
            return sqla_column_type.compile().upper()
        else:
            return super().column_datatype_to_string(sqla_column_type, dialect)

    @staticmethod
    def alias_in_select_mutator(label: str) -> str:
        return f"{label}__"

    @classmethod
    def get_timezone_expressions(
        cls,
        col: ColumnClause,
        timezone: typing.Optional[str],
    ) -> ColumnClause:
        if BF_EUROPE_BERLIN_TIMEZONE_COLUMN_SUFFIX in col.name:
            return col
        tpl = f"""
CAST(
    FROMDATETIME(
        DATETIMECONVERT(
            {col},
            '1:MILLISECONDS:EPOCH',
            '1:MILLISECONDS:SIMPLE_DATE_FORMAT:yyyy-MM-dd HH:mm:ss.SSS tz({timezone})',
            '1:MILLISECONDS'
        ),
        'yyyy-MM-dd HH:mm:ss.SSS'
    ) AS TIMESTAMP
)""".strip()
        return literal_column(tpl)
```

Reply:
@zhaoyongjie I'm aware of what the Python Date Format (PDF) represents, though thanks for clarifying that this shouldn't be used for ad-hoc expressions.

Note we [do already have](https://github.com/apache/superset/blob/master/superset/db_engine_specs/pinot.py) a Pino DB engine spec, but maybe only adding the `column_datatype_to_string` method is required.

Reply:
The Pinot provided time grain function so we just define `_time_grain_expressions` and apply `column_datatype_to_string` is enough.

Reply:
@zhaoyongjie do we just update this code to use the driver at https://github.com/BurdaForward/pinot-dbapi/tree/bf_release and that will allow ad-hoc expressions to work with Pinot?  Or do I change the existing driver to follow the code snippet you provided above?

Reply:
Hi @ege-st I think you only replace content of the `pinot.py` with codes snippet above which is to resolve this PR issue. The owner custom driver is for other issue, e.g. connection timeout and so on.

Reply:
@zhaoyongjie Thanks, I will test out your recommendation today and update the PR accordingly.

Reply:
@zhaoyongjie after some testing there's one problem that's preventing your suggestion from working with Pinot. When using the code above, it will generate a query that looks like this:
```
SELECT CAST(DATE_TRUNC('day', CAST(created_date_ts AS TIMESTAMP)) AS TIMESTAMP) as created_date_ts,
       COUNT(*) AS "count"
FROM "default".test_data
GROUP BY CAST(DATE_TRUNC('day', CAST(created_date_ts AS TIMESTAMP)) AS TIMESTAMP)
ORDER BY COUNT(*) DESC
LIMIT 10000;
```
Unfortunately, if you alias the expression as the name of an already existing column, Pinot will not properly analyze the `GROUP BY` clause.  The above will generate an error.

However, if I change the query to:
```
SELECT CAST(DATE_TRUNC('day', CAST(created_date_ts AS TIMESTAMP)) AS TIMESTAMP) as created_date_ts_2,
       COUNT(*) AS "count"
FROM "default".test_data
GROUP BY CAST(DATE_TRUNC('day', CAST(created_date_ts AS TIMESTAMP)) AS TIMESTAMP)
ORDER BY COUNT(*) DESC
LIMIT 10000;
```
or if I use:
```
SELECT CAST(DATE_TRUNC('day', CAST(created_date_ts AS TIMESTAMP)) AS TIMESTAMP) as created_date_ts,
       COUNT(*) AS "count"
FROM "default".test_data
GROUP BY created_date_ts
ORDER BY COUNT(*) DESC
LIMIT 10000;
```

It will work. Is there a way that I can change what alias Superset uses for the computed timestamp?

Reply:
@ege-st I might be a potential bug in Apache Calcite that was the SQL parser in Pinot. It was not too hard to fix it in Pinot codebase, I remember this issue was existing some DB which is used Calcite as parser.

Another option is add changes in Superset codes.

```python
diff --git a/superset/connectors/sqla/models.py b/superset/connectors/sqla/models.py
index 5c5f66853..1fcdff8e5 100644
--- a/superset/connectors/sqla/models.py
+++ b/superset/connectors/sqla/models.py
@@ -1081,7 +1081,7 @@ class SqlaTable(Model, BaseDatasource):  # pylint: disable=too-many-public-metho
         # add quotes to tables
         if db_engine_spec.allows_alias_in_select:
             label = db_engine_spec.make_label_compatible(label_expected)
-            sqla_col = sqla_col.label(label)
+            sqla_col = sqla_col.label(db_engine_spec.alias_in_select_mutator(label))
         sqla_col.key = label_expected
         return sqla_col

diff --git a/superset/db_engine_specs/base.py b/superset/db_engine_specs/base.py
index df0552e8a..1d8e91482 100644
--- a/superset/db_engine_specs/base.py
+++ b/superset/db_engine_specs/base.py
@@ -644,6 +644,10 @@ class BaseEngineSpec:  # pylint: disable=too-many-public-methods
             )
         )

+    @staticmethod
+    def alias_in_select_mutator(label: str) -> str:
+        return label
+
     @classmethod
     def fetch_data(
         cls, cursor: Any, limit: Optional[int] = None
diff --git a/superset/db_engine_specs/pinot.py b/superset/db_engine_specs/pinot.py
index d88fdf36f..d82bf8604 100644
--- a/superset/db_engine_specs/pinot.py
+++ b/superset/db_engine_specs/pinot.py
@@ -26,7 +26,7 @@ class PinotEngineSpec(BaseEngineSpec):  # pylint: disable=abstract-method
     engine_name = "Apache Pinot"
     allows_subqueries = False
     allows_joins = False
-    allows_alias_in_select = False
+    allows_alias_in_select = True
     allows_alias_in_orderby = False

     # https://docs.pinot.apache.org/users/user-guide-query/supported-transformations#datetime-functions
@@ -56,3 +56,7 @@ class PinotEngineSpec(BaseEngineSpec):  # pylint: disable=abstract-method
             return sqla_column_type.compile().upper()
         else:
             return super().column_datatype_to_string(sqla_column_type, dialect)
+
+    @staticmethod
+    def alias_in_select_mutator(label: str) -> str:
+        return f"{label}__"

```

Reply:
BTW, before making this change, also please check the latest Pinot that whether or not fix it in Pinot side.

Reply:
@zhaoyongjie agreed, I started looking to see if this is a bug in Pinot or in Calcite but I also want to see if I can get the fix done entirely within Superset because upgrading the Superset version will be easier than upgrading the Pinot deployment and, just in case, the fix in Calcite or Pinot takes a while to get done. 

I'll take a look at the Superset changes you suggested.

Reply:
@ege-st I also wondered if this was an underlying issue with the Pino SQLAlchemy dialect. You might want to look into the `visit_label` method.

Reply:
@zhaoyongjie I've confirmed that this error happens with the latest versions of Pinot: so Superset can't alias a projection to the same name as a column that already exists.  I looked at the diff you provided but it appears to be diffing a version of `models.py` that is not the same as the one in the `master` branch.

@john-bodley could you provide some more detail? Is SQL Alchemy generating the alias name used in the projection? If so, then it could be an issue with the dialect, but if Superset generates the alias label then I'm not sure how the dialect can address this.

Reply:
@zhaoyongjie so I believe I figured out a workaround for the alias issue.  If I just set `allows_alias_in_select = False` then the query generated by Superset does not use an alias and the query is then compatible with Pinot. So, I don't think any of the additional changes you kindly suggested are necessary.

One question that I have is: what is the purpose of the `pdf` that gets defined in the dataset configuration?  Since it isn't passed into the engine spec when creating a chart, it can't be used in the query generation, so it doesn't seem to serve a purpose?

Reply:
@ege-st 
> so I believe I figured out a workaround for the alias issue. If I just set allows_alias_in_select = False then the query generated by Superset does not use an alias and the query is then compatible with Pinot. So, I don't think any of the additional changes you kindly suggested are necessary.

Sounds good! It should be worked. 

> One question that I have is: what is the purpose of the pdf that gets defined in the dataset configuration? Since it isn't passed into the engine spec when creating a chart, it can't be used in the query generation, so it doesn't seem to serve a purpose?

I think the original design of "pdf" is a hard-code for getting a timestamp from a string, but a type conversion expression is more graceful.